DM-16648: new classes for DataUnits (now Dimensions)

TallJimbo opened the pull request:
This is one piece of DM-15034, but as it's a pretty big redesign I wanted to get some early feedback.

This is hence a PR against tickets/DM-15034, not master, though those are currently the same.  I'll merge this into that branch when it's reviewed, and then merge other sub-task tickets to that branch until all of DM-15034 is done.  That will include making all of the other Registry and Preflight code use these classes and removing the old ones, as well as adding some new classes to represent Data IDs in at least some contexts.

TallJimbo committed:
Add new classes for DataUnit system (now called Dimensions).

This is not yet being used anywhere; it exists alongside the
old implementation.

TallJimbo committed:
Add utility metaclass for hiding constructors.

andy-slac commented on config/dimensions.yaml:
```diff
@@ -0,0 +1,260 @@
+
+    VisitDetectorSkyPixJoin:
+      doc: >
+        A hasRegion join table that relates Visit+Detector to SkyPix, also used
```
Not very clear from the doc itself what "hasRegion" join table means here.

TallJimbo replied:
Unintentional, overzealous find-and-replace from "spatial" to "hasRegion"; will fix.

andy-slac commented on python/lsst/daf/butler/core/dimensions/elements.py:
```diff
@@ -0,0 +1,194 @@
+
+
+class DimensionUnit(Dimension):
+    r"""A discrete unit of data used to organize `Dataset`\s and associate
```
I think we had trouble with flake8 and raw docstrings in the past, to avoid backslashes references can be written as `` `Datasets <Dataset>` ``

TallJimbo replied:
I think I once had a similar conversation with @timj and @jonathansick that landed on the opposite conclusion, with raw docstrings preferred.  I hope one of them can chime in with a clarification.

andy-slac commented on python/lsst/daf/butler/core/dimensions/elements.py:
```diff
@@ -0,0 +1,194 @@
+        return self._allDependencies if optional else self._requiredDependencies
+
+
+class DimensionUnit(Dimension):
```
`DimensionUnit` confuses me somewhat (not sure about other people). Would it be clearer to just call this class `Dimension` (and base class something else)?

TallJimbo replied:
Agreed.

TallJimbo replied:
I gave this a try, and it doesn't quite work; the names for lots of other attributes and methods make more sense if the name for the base concept is really a first-class name.

Would `elements` instead of `units` be any better?  Any other ideas?

andy-slac commented on python/lsst/daf/butler/core/dimensions/elements.py:
```diff
@@ -0,0 +1,194 @@
+            return NotImplemented
+
+    def __hash__(self):
+        return hash(tuple(self.name))
```
`tuple(self.name, )` or `hash(self.name)`?

TallJimbo replied:
Good catch, should have been the latter.

andy-slac commented on python/lsst/daf/butler/core/dimensions/elements.py:
```diff
@@ -0,0 +1,194 @@
+    @property
+    def primaryKey(self):
+        """The names of fields that uniquely identify this dimension in a
+        data ID dict(`frozenset` of `str`).
```
Space before paren.

andy-slac commented on python/lsst/daf/butler/core/dimensions/elements.py:
```diff
@@ -0,0 +1,194 @@
+        (`DimensionUnitSet`).
+
+        Left vs. right is completely arbitrary; the terminology simply provides
+        an easy way to distiguish between the two sides.
```
Should we drop lhs/rhs entirely if they are not useful and just do `def joins(): return (lhs, rhs)`?

TallJimbo replied:
Possibly; I'd like to try to go through the transition to using these classes first to verify that we don't need them first.

andy-slac commented on python/lsst/daf/butler/core/dimensions/elements.py:
```diff
@@ -0,0 +1,194 @@
+    @property
+    def summarizes(self):
+        r"""A set of other `DimensionJoin`\s that provide more fine-grained
+        relationships this one (`DimensionJoinSet`).
```
_than_ this one?

andy-slac commented on python/lsst/daf/butler/core/dimensions/sets.py:
```diff
@@ -0,0 +1,402 @@
+    """
+
+    def __repr__(self):
+        return f"DimensionUnitSet({str(self)})"
```
Do you need `str` here, I think it's implied?

TallJimbo replied:
I wasn't sure it was `str` that was implied rather than `repr`, which would have been bad.  But it looks like you're right.

andy-slac commented on python/lsst/daf/butler/core/dimensions/sets.py:
```diff
@@ -0,0 +1,402 @@
+        expanded = set()
+
+        # recursive local function to add dependencies of dependencies
+        def visit(unit):
```
I'd probably use generator instead of closure for this.

TallJimbo replied:
When you get a chance, could you elaborate?  I'm not quite sure what you're proposing.

andy-slac replied:
I mean something like:


            def visit(unit):
                yield unit.name
                for dep in unit.dependencies(optional=optional):
                    yield from visit(dep)

            expaned = set(visit(unit))

andy-slac commented on python/lsst/daf/butler/core/dimensions/sets.py:
```diff
@@ -0,0 +1,402 @@
+        for dimension in self.universe._joins:
+            if dimension.name in names:
+                self._elements[dimension.name] = dimension
+                names.remove(dimension.name)
```
can `names` be a `frozenset`?

TallJimbo replied:
Apparently not; I'll transform to `set` at the beginning of this function both to ensure it is one and avoid argument-mutating side-effects.

andy-slac reviewed (commented):
I only looked at few modules and they look OK (but there are many new things so it needs some guesswork), few comments for now. I'll try to look at the rest later today if I can find time.

andy-slac commented on python/lsst/daf/butler/core/dimensions/elements.py:
```diff
@@ -0,0 +1,194 @@
+from ..utils import PrivateConstructorMeta
+from .sets import DimensionUnitSet
+
+__all__ = ("Dimension", "DimensionUnit", "DimensionJoin")
```
This is supposed to be before imports or it may not work as expected (same for other files).

andy-slac commented on python/lsst/daf/butler/core/dimensions/sets.py:
```diff
@@ -0,0 +1,402 @@
+
+class UniverseMismatchError(Exception):
+    """Exception thrown when trying to relate dimension objects from one
+    universe with those from another.
```
So `daf_butler` is ready to be deployed across multiverse now? 😃

andy-slac commented on python/lsst/daf/butler/core/dimensions/graph.py:
```diff
@@ -0,0 +1,362 @@
+     of all of its elements.
+
+     - `joins()` returns all joins whose dependencies are present in `units`
+     (though by default it may filter out some of these; see the `join`
```
`` `join` `` -> `` `joins` ``

andy-slac commented on python/lsst/daf/butler/core/dimensions/graph.py:
```diff
@@ -0,0 +1,362 @@
+    `DimensionGraph`\s can only be constructed by loading definitions from
+    config via the `load` method.  This returns a special "universe"
+    `DimensionGraph` that defines the maximal `DimensionUnit` and
+    `DimensionJoin` sets.  Dimension objects from different objects are not
```
"different objects" means different graphs?

TallJimbo replied:
Different universes; you can create one universe graph and call `subgraph` to make a new graph that is interoperable with the first.

andy-slac commented on python/lsst/daf/butler/core/dimensions/graph.py:
```diff
@@ -0,0 +1,362 @@
+            # dependencies are already in the universe.
+            namesOfUnblockedUnits = [name for name, deps in unitsToDo.items()
+                                     if deps.issubset(universe._units.names)]
+            namesOfUnblockedUnits.sort()  # break topological ties with lexographical sort
```
lexographical -> lexicographical  (below too)

andy-slac commented on python/lsst/daf/butler/core/dimensions/graph.py:
```diff
@@ -0,0 +1,362 @@
+
+        # The 'summarizes' field of DimensionJoin needs to be populated in the
+        # opposite order (DimensionJoins that relate more DimensionUnits
+        # summarize those that relate fewer), so we put off setting those
```
I think it's the opposite, fewer Dimensions should summarize more Dimensions?

andy-slac commented on python/lsst/daf/butler/core/dimensions/graph.py:
```diff
@@ -0,0 +1,362 @@
+            if not isinstance(join, DimensionJoin):
+                join = self._joins[join]
+            else:
+                UniverseMismatchError.check(self, unit)
```
`unit` -> `join`?

andy-slac reviewed (approved):
Few more (silly) comments. Looks fine, don't have time to look at unit test though, presume it's fine too.

TallJimbo committed:
Big refactor of Dimensions classes, to evaluate before squashing.

 - Renamed DimensionUnit -> Dimension, Dimension->DimensionElement
 - Just one DimensionSet class.
 - DimensionGraph now iterates over Dimensions.
 - DimensionSet and DimensionGraph are interoperable with regular
 sets and other containers, including containers of names.
 - Standard construction pattern is now to intersect a graph or set
 with a set of names.

 Still need to improve docstrings after this change, but I'm deferring
 that until I've tried using it in practice a bit.

TallJimbo commented:
While this initial review was very helpful, the design has moved on quite a bit, and the integration with the rest of daf_butler is now complete.  I'm just going to close this PR and open a new one for the entirety of DM-15034.
